src/server/pfs/server/driver_fsck.go

```diff
@@ -125,172 +125,237 @@ func (e ErrCommitAncestryBroken) Error() string {
+	}
 
-	// collect all the info for the branches and commits in pfs
-	branchInfos := make(map[string]*pfs.BranchInfo)
-	commitInfos := make(map[string]*pfs.CommitInfo)
-	newCommitInfos := make(map[string]*pfs.CommitInfo)
-	repoInfo := &pfs.RepoInfo{}
-	if err := d.repos.ReadOnly(ctx).List(repoInfo, col.DefaultOptions(), func(string) error {
-		commitInfo := &pfs.CommitInfo{}
-		if err := d.commits.ReadOnly(ctx).GetByIndex(pfsdb.CommitsRepoIndex, pfsdb.RepoKey(repoInfo.Repo), commitInfo, col.DefaultOptions(), func(string) error {
-			commitInfos[pfsdb.CommitKey(commitInfo.Commit)] = proto.Clone(commitInfo).(*pfs.CommitInfo)
-			return nil
+	if !equalBranches(append(bi.Provenance, bi.Branch), followedProvenances) {
+		if err := onError(ErrBranchProvenanceTransitivity{
+			BranchInfo:     bi,
+			FullProvenance: followedProvenances,
 		}); err != nil {
-			return errors.EnsureStack(err)
+			return err
 		}
-		branchInfo := &pfs.BranchInfo{}
-		err := d.branches.ReadOnly(ctx).GetByIndex(pfsdb.BranchesRepoIndex, pfsdb.RepoKey(repoInfo.Repo), branchInfo, col.DefaultOptions(), func(string) error {
-			branchInfos[pfsdb.BranchKey(branchInfo.Branch)] = proto.Clone(branchInfo).(*pfs.BranchInfo)
-			return nil
-		})
-		return errors.EnsureStack(err)
-	}); err != nil {
-		return errors.EnsureStack(err)
 	}
 
-	// for each branch
-	for _, bi := range branchInfos {
-		// we expect the branch's provenance to equal the union of the provenances of the branch's direct provenances
-		// i.e. union(branch, branch.Provenance) = union(branch, branch.DirectProvenance, branch.DirectProvenance.Provenance)
-		direct := bi.DirectProvenance
-		union := []*pfs.Branch{bi.Branch}
-		for _, directProvenance := range direct {
-			directProvenanceInfo := branchInfos[pfsdb.BranchKey(directProvenance)]
-			union = append(union, directProvenance)
-			if directProvenanceInfo != nil {
-				union = append(union, directProvenanceInfo.Provenance...)
+	return nil
+}
+
+func checkBranchSubvenances(bi *pfs.BranchInfo, branchInfos map[string]*pfs.BranchInfo, onError func(error) error) error {
+	for _, provBranch := range bi.Provenance {
+		provBranchInfo := branchInfos[pfsdb.BranchKey(provBranch)]
+		if !branchInSet(bi.Branch, provBranchInfo.Subvenance) {
+			if err := onError(ErrBranchSubvenanceTransitivity{
+				BranchInfo:        provBranchInfo,
+				MissingSubvenance: bi.Branch,
+			}); err != nil {
+				return err
 			}
 		}
+	}
+
+	return nil
+}
 
-		if !equalBranches(append(bi.Provenance, bi.Branch), union) {
-			if err := onError(ErrBranchProvenanceTransitivity{
-				BranchInfo:     bi,
-				FullProvenance: union,
+func checkBranchCommitProvenance(bi *pfs.BranchInfo, branchInfos map[string]*pfs.BranchInfo, commitInfos map[string]*pfs.CommitInfo, onError func(error) error) error {
+	// build set of all commits in branch head direct provenance
+	headProvenance := make(map[string]bool)
+	headCI, ok := commitInfos[bi.Head.ID]
```

Comment:
this should be `headCI, ok := commitInfos[pfsdb.CommitKey(bi.Head)]`

Reply:
fixed